### ISSUE TITLE
Compilation Error: secp256k1-sys fails to build for wasm32-wasi target

### ISSUE THREAD

**Issue by Th0rgal** [rust-bitcoin/rust-bitcoin#3447]
I'm working on a Rust project that targets wasm32-wasi to generate WebAssembly that can be used in JavaScript. My goal is to recompute the Merkle root of a block from transactions, and I do not require signing capabilities. However, I'm encountering issues with the secp256k1-sys dependency when building for this target.

I modified my Cargo.toml as follows:
```toml
   [package]
   name = "utu_core"
   version = "0.1.0"
   edition = "2021"

   [dependencies]
   bitcoin = { version = "0.32.3", default-features = false, features = ["no_std"] }

   [lib]
   crate-type = ["cdylib"]
   ```
   
   I try building my project this way:
   ```
      cargo wasi build -p utu_core
   ```
   
   and got this error:
   ```
   Compiling secp256k1-sys v0.10.1
The following warnings were emitted during compilation:

warning: secp256k1-sys@0.10.1: error: unable to create target: 'No available targets are compatible with triple "wasm32-unknown-wasi"'
warning: secp256k1-sys@0.10.1: 1 error generated.
...

error: failed to run custom build command for `secp256k1-sys v0.10.1`

Caused by:
  process didn't exit successfully: `/path/to/build-script-build` (exit status: 1)
  --- stdout
  ...
  cargo:warning=error: unable to create target: 'No available targets are compatible with triple "wasm32-unknown-wasi"'
  cargo:warning=1 error generated.
  ...
  ```
  
  Is there a way to bypass the secp256k1-sys dependency or configure it to work with wasm32-wasi?

**Comment by apoelstra** [comment]
Unfortunately not now. We are working toward this as part of our broader "crate smashing" exercise to thoroughly separate out our dependencies.

cc #526

**Comment by imikushin** [comment]
@apoelstra Would you guys consider accepting a PR replacing secp256k1 with a pure Rust alternative? Looks like there are some decent options out there: `k256`, `ark-secp256k1` — to take a couple.

I'd be honored to contribute and introduce rust-bitcoin to the Wasm crowd.

**Comment by apoelstra** [comment]
@imikushin absolutely not. No pure Rust implementation has anything approaching the QA cycles and the anti-sidechannel protections of the libsecp256k1 C library.

**Comment by imikushin** [comment]
@apoelstra fair. Perhaps, replacing is too strong of an ask. How about adding an optional feature to secp256k1 crate swapping out the libsecp256k1 C implementation with k256? 

Anything touching a private key could even be stubbed out with `unimplemented!()`.

The idea is to be able to read Bitcoin transactions and verify signatures in Wasm, nothing more.

**Comment by apoelstra** [comment]
Yes, we would like to do parsing and serialization in pure Rust eventually. We have most of the code here https://github.com/rust-bitcoin/rust-secp256k1/pull/703 though it needs to be rebased and reviewed.

With all due respect to Tony and the rust-crypto team, we have had many differences of opinion over the years and I do not feel comfortable taking a dependency on their crate. Especially on a crate that implements a bunch of pure-Rust signing logic which would make our users uncomfortable, even if we promised that it was disabled and that no secret data ever crossed their API boundary.

Regardless of what promises we could make, we would need to review k256 with the same rigor as we would review #703, and that review time would be better spent on an incremental, definitely-non-signing, designed-to-be-compatible-with-libsecp256k1, series of PRs that we could maintain going forward.

**Comment by imikushin** [comment]
@apoelstra Totally get it! I'll check out https://github.com/rust-bitcoin/rust-secp256k1/pull/703 

My whole premise even beginning to look into this direction started with me having a hard time compiling rust-bitcoin and rust-secp256k1 to Wasm. 

Luckily, I have just discovered, that both crates compile to Wasm quite nicely if one does the pre-requisite work: install LLVM and add it to PATH 😅

I'd be delighted to contribute to documenting these steps for future users.

**Comment by apoelstra** [comment]
Cool! Yes, we would greatly appreciate documentation efforts on this front. (Especially if you have the process nailed down enough to write a script we can put in our CI.)

**Comment by tcharding** [comment]
We also have a cookbok at https://rust-bitcoin.org (repo: https://github.com/rust-bitcoin/rust-bitcoin.github.io) that could have a WASM section in it if you are super keen.

**Comment by Kixunil** [comment]
FTR, the current limitations of `cargo` would make such replacing very annoying anyway. If you need signature verification and signing, you could just write a clib with API compatible with `libsecp256k1`. I discussed a similar idea recently in the secp repo.

(Note: the OP is confused - WASI and JS are two different things.)
